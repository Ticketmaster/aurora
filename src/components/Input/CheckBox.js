import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import classNames from "classnames";
import Checkmark from "../Icons/Checkmark";
import { Consumer } from "./Context";
import constants from "../../theme/constants";
import colors from "../../theme/colors";
import composeEventHandlers from "../../utils/composeEventHandlers";

const CheckBoxInput = styled.input.attrs({
  type: "checkbox"
})`
  outline: none;

  margin: 0;
  appearance: none;
  position: relative;
  .checkbox--large & {
    width: 24px;
    height: 24px;
  }
  .checkbox--small & {
    width: 16px;
    height: 16px;
  }

  &:before {
    content: "";
    background-color: transparent;
    border-radius: ${constants.borderRadius.small};
    border: 1px solid ${colors.onyx.light};
    position: absolute;
    .checkbox--large & {
      width: 24px;
      height: 24px;
    }
    .checkbox--small & {
      width: 16px;
      height: 16px;
    }
    &.checkbox--disabled {
      color: grey;
    }
  }
  &:hover:before {
    border-width: 2px;
    border-color: ${colors.azure.base};
  }
  &:focus:before {
    .checkbox--small & {
      width: 16px;
      height: 16px;
      border-radius: ${constants.borderRadius.small};
      border: solid 1px ${colors.azure.base};
      background-color: ${colors.white.base};
    }
    .checkbox--large & {
      width: 24px;
      height: 24px;
      border-radius: ${constants.borderRadius.small};
      border: solid 1px ${colors.azure.base};
      background-color: ${colors.white.base};
    }
  }
  &:checked:after {
    content: "";
    position: absolute;
    .checkbox__multi--large & {
      content: "";
      width: 12px;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      height: 3px;
      border-radius: ${constants.borderRadius.small};
      background-color: ${colors.azure.base};
    }
    .checkbox__multi--small & {
      content: "";
      width: 8px;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      height: 2px;
      border-radius: ${constants.borderRadius.small};
      background-color: ${colors.azure.base};
    }
  }
`;

const CheckBoxWrapper = styled.label`
  position: relative;
  cursor: pointer;
  display: flex;
  align-items: center;
  margin-bottom: 10px;
  text-align: left;
  color: ${colors.onyx.base};
  &.checkbox--disabled {
    color: grey;
    cursor: default;
    pointer-events: none;
  }
  &.checkbox--large {
    font-size: 16px;
    line-height: 1.5;
  }
  &.checkbox--small {
    font-size: 14px;
    line-height: 1.5;
  }
`;

const CheckBoxText = styled.span`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: 7px;
`;

const CheckDivStyling = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: left;
  position: absolute;
  opacity: 0;

  .checkbox__checked & {
    opacity: 1;
  }
  .checkbox__multi--small & {
    display: none;
  }

  .checkbox__multi--large & {
    display: none;
  }
`;

const StyledCheckmark = styled(Checkmark)`
  margin-top: 4px;
`;

const CheckBoxButton = ({
  name,
  value,
  size,
  isActive,
  children,
  ...props
}) => (
  <Consumer>
    {({ selected, onClick }) => {
      const isChecked = selected.includes(value);
      return (
        <CheckBoxWrapper
          className={classNames({
            "checkbox--small": size === "small",
            "checkbox--large": size === "large",
            "checkbox--disabled": !isActive,
            "checkbox__multi--large": selected.length > 1 && size === "large",
            "checkbox__multi--small": selected.length > 1 && size === "small",
            checkbox__checked: isChecked
          })}
        >
          <CheckBoxInput
            {...props}
            value={value}
            name={name}
            id={name + value}
            disabled={!isActive}
            size={size}
            defaultChecked={isChecked}
            onClick={composeEventHandlers(
              () => onClick({ value }),
              props.onClick
            )}
          />
          <CheckDivStyling
            className={classNames({
              checkbox__checked: isChecked
            })}
          >
            <StyledCheckmark size={size === "small" ? 16 : 24} />
          </CheckDivStyling>
          <CheckBoxText
            role="checkbox"
            aria-labelledby={name + value}
            aria-checked={isChecked}
          >
            {children}
          </CheckBoxText>
        </CheckBoxWrapper>
      );
    }}
  </Consumer>
);

CheckBoxButton.propTypes = {
  size: PropTypes.oneOf(["large", "small"]),
  children: PropTypes.node,
  isActive: PropTypes.bool,
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  multiSelect: PropTypes.bool,
  onClick: PropTypes.func
};

CheckBoxButton.defaultProps = {
  size: "small",
  children: null,
  isActive: true,
  multiSelect: false,
  onClick: () => {}
};

export default CheckBoxButton;
