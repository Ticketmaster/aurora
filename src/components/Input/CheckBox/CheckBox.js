import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { withTheme } from "styled-components";

import composeEventHandlers from "../../../utils/composeEventHandlers";
import {
  CheckBoxWrapper,
  CheckBoxText,
  CheckBoxInput,
  CheckDivStyling,
  StyledCheckmark
} from "./CheckBox.styles";
import { themes } from "../../../theme";

export const blur = e => {
  e.target.blur();
}; // to avoid focus after click

const CheckBoxButton = ({
  name,
  value,
  size,
  disabled,
  children,
  checked,
  style,
  onClick,
  theme: { themeName },
  ...props
}) => {
  const id = (name + value).replace(/ /g, "_");
  const isSmall = size === "small";
  const inputProps = {
    ...props,
    value,
    name,
    disabled,
    ...(checked !== null && { checked })
  };

  return (
    <CheckBoxWrapper
      className={classNames({
        "checkbox--small": isSmall,
        "checkbox--large": !isSmall,
        "checkbox--disabled": disabled,
        checkbox__checked: checked,
        checkbox__indeterminate: checked === "indeterminate"
      })}
      style={style}
    >
      <CheckBoxInput
        aria-labelledby={id}
        {...inputProps}
        onClick={composeEventHandlers(blur, onClick)}
      />
      <CheckDivStyling
        className={classNames({
          checkbox__checked: checked
        })}
      >
        <StyledCheckmark
          size={isSmall ? 16 : 24}
          stroke={
            disabled
              ? themes[themeName].onyx.muted
              : themes[themeName].primary.base
          }
        />
      </CheckDivStyling>
      <CheckBoxText id={id}>{children}</CheckBoxText>
    </CheckBoxWrapper>
  );
};

CheckBoxButton.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  theme: PropTypes.shape({
    themeName: PropTypes.string
  }).isRequired,
  checked: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.oneOf(["indeterminate"])
  ]),
  size: PropTypes.oneOf(["large", "small"]),
  style: PropTypes.shape(),
  children: PropTypes.node,
  disabled: PropTypes.bool,
  onClick: PropTypes.func
};

CheckBoxButton.defaultProps = {
  size: "small",
  children: null,
  disabled: false,
  checked: null,
  onClick: null,
  style: {}
};

export default withTheme(CheckBoxButton);
