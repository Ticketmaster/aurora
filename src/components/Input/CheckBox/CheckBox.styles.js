import styled from "styled-components";
import Checkmark from "../../Icons/Checkmark";
import { typography } from "../../../theme";
import { getThemeValue } from "../../../utils";

export const CheckBoxWrapper = styled.label`
  position: relative;
  outline: none;
  cursor: pointer;
  display: flex;
  align-items: center;
  text-align: left;

  &.checkbox--disabled {
    opacity: 0.4;
    pointer-events: none;
  }
  &.checkbox--large {
    font-size: ${typography.size.kilo};
    line-height: 1.25;
  }
  &.checkbox--small {
    font-size: ${typography.size.hecto};
    line-height: 1.29;
  }
`;

export const CheckBoxText = styled.span`
  color: ${getThemeValue("gray01")};
  display: flex;
  outline: none;
  align-items: center;
  justify-content: flex-start;
  margin-left: 7px;
  width: 90%;
`;

export const CheckDivStyling = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: left;
  position: absolute;
  opacity: 0;
  color: ${getThemeValue("primary", "base")};

  .checkbox__checked & {
    opacity: 1;
  }
  .checkbox__indeterminate & {
    display: none;
  }
  .checkbox--disabled & {
    color: ${getThemeValue("gray02")};
  }
`;

export const StyledCheckmark = styled(Checkmark)`
  margin-top: 3px;
`;
