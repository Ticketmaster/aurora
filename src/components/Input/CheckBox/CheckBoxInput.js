import React from "react";
import PropTypes from "prop-types";
import CheckBoxInput from "./CheckBoxInput.styles";

class CheckBoxInputComponent extends React.Component {
  static propTypes = {
    size: PropTypes.oneOf(["large", "small"]),
    disabled: PropTypes.bool,
    name: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    isChecked: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
    onChange: PropTypes.func,
    isFocused: PropTypes.bool
  };

  static defaultProps = {
    size: "small",
    disabled: false,
    isFocused: false,
    onChange: () => null
  };

  constructor(props) {
    super(props);
    this.SelectedElement = React.createRef();
  }

  componentDidUpdate() {
    this.props.isFocused && this.SelectedElement.current.focus(); // eslint-disable-line
  }

  render() {
    const { name, value, size, disabled, isChecked, ...props } = this.props;
    return (
      <CheckBoxInput
        {...props}
        value={value}
        name={name}
        id={`${name + value}input`}
        disabled={disabled}
        checked={isChecked}
        innerRef={this.SelectedElement}
      />
    );
  }
}
export default CheckBoxInputComponent;
