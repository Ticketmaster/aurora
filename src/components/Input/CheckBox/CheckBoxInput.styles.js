import styled from "styled-components";
import { constants } from "../../../theme";
import { getThemeValue } from "../../../utils";

export default styled.input.attrs({
  type: "checkbox"
})`
  outline: none;
  margin: 0;
  appearance: none;
  position: relative;
  background-color: ${getThemeValue("white", "base")};
  border-color: ${getThemeValue("gray02")};
  border-radius: ${constants.borderRadius.small};
  display: flex;
  justify-content: center;
  align-items: center;

  &::-ms-check {
    display: none;
  }

  .checkbox--large & {
    width: 24px;
    height: 24px;
  }

  .checkbox--small & {
    width: 16px;
    height: 16px;
  }

  &:active {
    transform: scale(0.95, 0.95);
  }

  &:before {
    content: "";
    box-sizing: border-box;
    background-color: transparent;
    border-radius: ${constants.borderRadius.small};
    border: 1px solid ${getThemeValue("gray02")};
    position: absolute;
    cursor: pointer;
    transition: border-width 0.3s ${constants.easing.easeInOutQuad};
    transition: border-color 0.3s ${constants.easing.easeInOutQuad};

    .checkbox--large & {
      width: 24px;
      height: 24px;
    }
    .checkbox--small & {
      width: 16px;
      height: 16px;
    }
    &.checkbox--disabled {
      color: grey;
    }
  }
  &:hover:before {
    border-width: 2px;
    border-color: ${getThemeValue("primary", "base")};
  }
  &:focus:before {
    outline: none;
    box-shadow: 0 0 5px 0 ${getThemeValue("primary", "base")};
    border-color: ${getThemeValue("primary", "base")};

    .checkbox--small & {
      width: 16px;
      height: 16px;
    }
    .checkbox--large & {
      width: 24px;
      height: 24px;
    }
  }

  &::after {
    content: "";
    box-sizing: border-box;
    transform: scale(0.7, 0.7);
    border-radius: ${constants.borderRadius.small};
    background-color: ${getThemeValue("primary", "base")};
    cursor: pointer;
    transition: all 0.3s ${constants.easing.elastic};

    .checkbox--disabled & {
      background-color: ${getThemeValue("gray02")};
    }
    .checkbox--small.checkbox__indeterminate & {
      width: 8px;
      height: 2px;
      transform: scale(1, 1);
    }
    .checkbox--large.checkbox__indeterminate & {
      width: 12px;
      height: 2px;
      transform: scale(1, 1);
    }
  }
`;
