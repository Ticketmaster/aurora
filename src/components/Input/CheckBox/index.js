/* eslint import/prefer-default-export: off */
export { default as CheckBox } from "./CheckBox";
