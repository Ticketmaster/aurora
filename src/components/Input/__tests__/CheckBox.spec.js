import React from "react";
import { ThemeProvider } from "styled-components";
import { render, cleanup, Simulate } from "react-testing-library";

import { CheckBox } from "../";
import { blur } from "../CheckBox/CheckBox";

describe("CheckBox", () => {
  afterEach(cleanup);

  it("renders default checkbox", () => {
    const { container } = renderCheckBox();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders default checked checkbox", () => {
    const { container } = renderCheckBox({ checked: true, onChange: () => {} });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders default indeterminate checkbox", () => {
    const { container } = renderCheckBox({
      checked: "indeterminate",
      onChange: () => {}
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders large checkbox", () => {
    const { container } = renderCheckBox({ size: "large" });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders disabled checkbox", () => {
    const { container } = renderCheckBox({ disabled: true });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders correctly after click ", () => {
    const { container } = renderCheckBox();
    Simulate.click(container.querySelector("label"));
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders correctly on focus ", () => {
    const { container } = renderCheckBox();
    Simulate.focus(container.querySelector("label"));
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should execute blur method on event target", () => {
    const mockFn = jest.fn();
    const mockEventObj = {
      target: {
        blur: mockFn
      }
    };
    blur(mockEventObj);
    expect(mockFn.mock.calls.length).toBe(1);
  });

  function renderCheckBox(props = {}) {
    return render(
      <ThemeProvider theme={{ themeName: "tm" }}>
        <CheckBox name="test_name" value="test_value" {...props}>
          Test checkbox
        </CheckBox>
      </ThemeProvider>
    );
  }
});
