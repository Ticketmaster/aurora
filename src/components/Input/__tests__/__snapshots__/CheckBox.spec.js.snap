// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox renders correctly after click  1`] = `
<label
  class="checkbox--small sc-bZQynM dgNshy"
>
  <input
    aria-labelledby="test_nametest_value"
    class="sc-iwsKbI evTFzO"
    name="test_name"
    type="checkbox"
    value="test_value"
  />
  <div
    class="sc-htoDjs egyiwf"
  >
    <svg
      class="sc-dnqmqq cOstNk"
      fill="none"
      fill-rule="evenodd"
      height="16"
      stroke="#026cdf"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="3"
      viewBox="0 0 24 28"
      width="16"
    >
      <path
        d="M5.215 12.219l4.437 4.439 9.514-9.433"
      />
    </svg>
  </div>
  <span
    class="sc-gzVnrw kcmiZY"
    id="test_nametest_value"
  >
    Test checkbox
  </span>
</label>
`;

exports[`CheckBox renders correctly on focus  1`] = `
<label
  class="checkbox--small sc-bZQynM dgNshy"
>
  <input
    aria-labelledby="test_nametest_value"
    class="sc-iwsKbI evTFzO"
    name="test_name"
    type="checkbox"
    value="test_value"
  />
  <div
    class="sc-htoDjs egyiwf"
  >
    <svg
      class="sc-dnqmqq cOstNk"
      fill="none"
      fill-rule="evenodd"
      height="16"
      stroke="#026cdf"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="3"
      viewBox="0 0 24 28"
      width="16"
    >
      <path
        d="M5.215 12.219l4.437 4.439 9.514-9.433"
      />
    </svg>
  </div>
  <span
    class="sc-gzVnrw kcmiZY"
    id="test_nametest_value"
  >
    Test checkbox
  </span>
</label>
`;

exports[`CheckBox renders default checkbox 1`] = `
<label
  class="checkbox--small sc-bZQynM dgNshy"
>
  <input
    aria-labelledby="test_nametest_value"
    class="sc-iwsKbI evTFzO"
    name="test_name"
    type="checkbox"
    value="test_value"
  />
  <div
    class="sc-htoDjs egyiwf"
  >
    <svg
      class="sc-dnqmqq cOstNk"
      fill="none"
      fill-rule="evenodd"
      height="16"
      stroke="#026cdf"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="3"
      viewBox="0 0 24 28"
      width="16"
    >
      <path
        d="M5.215 12.219l4.437 4.439 9.514-9.433"
      />
    </svg>
  </div>
  <span
    class="sc-gzVnrw kcmiZY"
    id="test_nametest_value"
  >
    Test checkbox
  </span>
</label>
`;

exports[`CheckBox renders default checked checkbox 1`] = `
<label
  class="checkbox--small checkbox__checked sc-bZQynM dgNshy"
>
  <input
    aria-labelledby="test_nametest_value"
    class="sc-iwsKbI evTFzO"
    name="test_name"
    type="checkbox"
    value="test_value"
  />
  <div
    class="checkbox__checked sc-htoDjs egyiwf"
  >
    <svg
      class="sc-dnqmqq cOstNk"
      fill="none"
      fill-rule="evenodd"
      height="16"
      stroke="#026cdf"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="3"
      viewBox="0 0 24 28"
      width="16"
    >
      <path
        d="M5.215 12.219l4.437 4.439 9.514-9.433"
      />
    </svg>
  </div>
  <span
    class="sc-gzVnrw kcmiZY"
    id="test_nametest_value"
  >
    Test checkbox
  </span>
</label>
`;

exports[`CheckBox renders default indeterminate checkbox 1`] = `
<label
  class="checkbox--small checkbox__checked checkbox__indeterminate sc-bZQynM dgNshy"
>
  <input
    aria-labelledby="test_nametest_value"
    class="sc-iwsKbI evTFzO"
    name="test_name"
    type="checkbox"
    value="test_value"
  />
  <div
    class="checkbox__checked sc-htoDjs egyiwf"
  >
    <svg
      class="sc-dnqmqq cOstNk"
      fill="none"
      fill-rule="evenodd"
      height="16"
      stroke="#026cdf"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="3"
      viewBox="0 0 24 28"
      width="16"
    >
      <path
        d="M5.215 12.219l4.437 4.439 9.514-9.433"
      />
    </svg>
  </div>
  <span
    class="sc-gzVnrw kcmiZY"
    id="test_nametest_value"
  >
    Test checkbox
  </span>
</label>
`;

exports[`CheckBox renders disabled checkbox 1`] = `
<label
  class="checkbox--small checkbox--disabled sc-bZQynM dgNshy"
>
  <input
    aria-labelledby="test_nametest_value"
    class="sc-iwsKbI evTFzO"
    disabled=""
    name="test_name"
    type="checkbox"
    value="test_value"
  />
  <div
    class="sc-htoDjs egyiwf"
  >
    <svg
      class="sc-dnqmqq cOstNk"
      fill="none"
      fill-rule="evenodd"
      height="16"
      stroke="rgba(38, 38, 38, 0.4)"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="3"
      viewBox="0 0 24 28"
      width="16"
    >
      <path
        d="M5.215 12.219l4.437 4.439 9.514-9.433"
      />
    </svg>
  </div>
  <span
    class="sc-gzVnrw kcmiZY"
    id="test_nametest_value"
  >
    Test checkbox
  </span>
</label>
`;

exports[`CheckBox renders large checkbox 1`] = `
<label
  class="checkbox--large sc-bZQynM dgNshy"
>
  <input
    aria-labelledby="test_nametest_value"
    class="sc-iwsKbI evTFzO"
    name="test_name"
    type="checkbox"
    value="test_value"
  />
  <div
    class="sc-htoDjs egyiwf"
  >
    <svg
      class="sc-dnqmqq cOstNk"
      fill="none"
      fill-rule="evenodd"
      height="24"
      stroke="#026cdf"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="3"
      viewBox="0 0 24 28"
      width="24"
    >
      <path
        d="M5.215 12.219l4.437 4.439 9.514-9.433"
      />
    </svg>
  </div>
  <span
    class="sc-gzVnrw kcmiZY"
    id="test_nametest_value"
  >
    Test checkbox
  </span>
</label>
`;
